### PR TITLE
Mods: Improve mod loading priorities (2nd)

### LINKED ISSUE #8601: Optional dependencies being overridden/turned into hard dependencies
This issue is very similar to #3136
(very nearly a duplicate, in fact, but that one's old, and long since closed)

Using 5.0.1-release, singleplayer

Consider this example dependency tree:

Mod A hard depends on mod C.  
B hard depends on D, optional-depends on F
C depends on nothing
D optional-depends on E
E hard depends on F

Now disable mod F.  Let's say you're doing this to test a feature in B that changes behavior depending on the presence of F.

In this case, E of course doesn't load, but D won't load either, because now the engine treats it as if it *hard-depends* on E, contrary to its `depends.txt`/`mod.conf`.  This causes B not to load as well, since it needs D.

So while one expects A, B, C, and D to load in this example, only A and C get loaded, thus preventing the modder from doing the test they wanted (to check B's behavior without F).

OK, now try it in a real setup:

* Download Dreambuilder
* Create a new world,
* Enable the whole modpack in that world's config
* Now disable just `mesecons` (the one base mod from that project), leaving everything else enabled
* Start the world.
* Many mods that should have loaded won't, for example `homedecor_lighting` (it has a number of dependencies, but mesecons is strictly optional)

Now, go back into the world config and disable all of the `mesecons*` mods (there are about 30, all grouped together), then save and start the world.  You'll find that many of the items that should have loaded the first time work now, even though you were turning things *off* in the config.

It's as though hard dependencies propagate upwards through the dependency tree once a mod can't be satisfied, overriding cases where mods are explicitly set to be optional.

### PR DESCRIPTION
**Pull request by SmallJoker**
Revival of #8603 because GitHub does not want to track branches that were `git reset --hard`'d.

Fixes #8601 and #9977 **but** breaks mods that rely on the current, undefined mod loading order. Please confirm that resulting mod errors are not caused by the different loading order.

## How it works
_Documentation in the code™_
There is now a new list to determinate how many mods depend on the mod:
   * deps_chain: Contains all mod names which either need or may need the mod

1) Dependencies are read from the mod's config
2) Each mod adds its own name recursively to the dependency list of each dependent mod
3) If a hard-dependency cannot be matched, the mod is not loaded (same as before)
4) If a recursion is detected, all affected mods will be loaded as long it's logically possible (i.e. enough optional dependencies)
5) Sort the final list: Sort the `ModSpec`s according to the amount of mods depending on this mod
6) Load the mods

## To do

Ready for review.

## How to test

1) Download [`mod_sorting_pack.zip`](https://github.com/minetest/minetest/files/3293490/mod_sorting_pack.zip)
2) Enable all mods inside the "Mod sorting modpack"
3) `debug_log_level = verbose`
4) Join a world
5) Confirm `Sorted mod: ` and `Loading and running script from` outputs (`grep` is your friend)
6) Test with your own complicated dependency issues, such as MineClone2 + unified_inventory
7) Run the unittests, because why not.

### CODE REVIEW

**Comment by debagos** [comment]
Hi, I hope this helps.
Here is what I did:
- Created a new fresh minetest from master, with #11438 applied
- Installed your `mod_sorting_pack` and additionally my two test mods from #9977
- Enabled `debug_log_level = verbose`
- Created a new world and joined

And this is what I got in my log:

- Some `Detected circular dependencies in mod XXXX` warnings
- Lots of `Sorted mod: XXXX depending=n` verbose messages
- And these two errors:
`mod "lm_f" has unsatisfied dependencies: `
`mod "lm_e" has unsatisfied dependencies:  "_foobar_not_a_mod_"`

This is not a Copy&Paste error, the `lm_f` mod error message really is like that in my log.

Good news is, this PR fixed #9977 and all unit tests passed.

**Comment by SmallJoker** [comment]
@debagos Thank you for testing. If #10486 works too, that one could be considered instead to have more detailed log messages. Although both PRs seem to do basically the same using slightly different approaches.

**Comment by Wuzzy2** [comment]
> Sort the final list: Sort the ModSpecs according to the amount of mods depending on this mod

But what was the reason for this choice of order? Seems rather unpredictable to me from the viewpoint of a single mod.

**Comment by SmallJoker** [comment]
> But what was the reason for this choice of order? Seems rather unpredictable to me from the viewpoint of a single mod.

It's the most logical order. Mods that are most important (for example `default`, MTG) are loaded first, and those which are not required by any other mod are loaded last. If you're relying on predictions, you definitely should properly configure the mod dependencies.
Mods that have equal priorities seem currently (master) to be sorted alphabetically. This might be different in this PR, but should totally not be a concern at all.

**Comment by rubenwardy** [comment]
Here's how I implemented mod ordering for my game: https://gitlab.com/rubenwardy/rvwp/-/blob/master/src/scripting/ModOrderResolver.cpp

The key here is to throw an error if the order can't be determined, rather than silently disabling mods. I think this is much more user friendly, and mod managing UIs should be able to help the user anyway

**Comment by Wuzzy2** [comment]
I think rubenwardy has a point. If a hard dependency cannot be resolved, don't start the game at all and just error out instead of forcefully trying to make it work *somehow*.

**Comment by SmallJoker** [comment]
> It's much simpler - just a depth-first search

Aborting right away would surely simplify it a bit, if players accept that change of behaviour. However, I don't see any way on how to reliably resolve optional dependencies there.

two cases
Left: question of what to pick first
Right: optional dependencies

**Comment by rubenwardy** [comment]
I don't see the problem - optional dependencies are handled the same as hard dependencies, you just don't throw an error if the mod doesn't exist. In case 1, they will be ordered correctly - as `CBA`.

My depth-first search is a simple solution if you're willing to accept an error rather than silently disabling mods

**Comment by wsor4035** [comment]
not trying to scope creep, however if the mod loading order is being improved, could https://github.com/minetest/minetest/issues/8934 be considered?

**Comment by SmallJoker** [comment]
@wsor4035 If rubenwardy's or appgurueu's suggestion is used, I don't know how that could be implemented. However with `std::sort` it would be possible to recursively add a z-index-alike value and check against that, rather than the amount of depending mods.

File: src/content/mods.cpp [hunk 1]
```diff
@@ -352,51 +352,132 @@ void ModConfiguration::checkConflictsAndDeps()
...
-void ModConfiguration::resolveDependencies()
+struct PrioritySortedMod
 {
-	// Step 1: Compile a list of the mod names we're working with
-	std::set<std::string> modnames;
-	for (const ModSpec &mod : m_unsatisfied_mods) {
-		modnames.insert(mod.name);
+	PrioritySortedMod(ModSpec *mod) : spec(mod) {}
+
+	static bool sorter(const PrioritySortedMod *a, const PrioritySortedMod *b)
```
Comment (sfan5) [inline src/content/mods.cpp:359]: If this was `operator<` couldn't `std::sort` make use of this automatically?

File: src/content/mods.cpp [hunk 2]
```diff
@@ -352,51 +352,132 @@ void ModConfiguration::checkConflictsAndDeps()
...
+{
+	if (deps_chain.find(modname) != deps_chain.end()) {
+		if (modname == spec->name) {
+			// Step 4: Reached the very same mod again.
+			// Stop the circular dependencies
+
+			// clang-format off
+			warningstream << "Detected circular dependencies in mod '" <<
+					modname << "'" << std::endl;
+			// clang-format on
```
Comment (sfan5) [inline src/content/mods.cpp:383]: clang-format comments are no longer needed

File: src/content/mods.cpp [hunk 3]
```diff
@@ -352,51 +352,132 @@ void ModConfiguration::checkConflictsAndDeps()
...
-		mod.unsatisfied_depends = mod.depends;
-		// check which optional dependencies actually exist
-		for (const std::string &optdep : mod.optdepends) {
-			if (modnames.count(optdep) != 0)
-				mod.unsatisfied_depends.insert(optdep);
+	bool addDependingMod(const std::string &modname,
+			std::unordered_map<std::string, PrioritySortedMod> &mods_by_name);
+
+	std::unordered_set<std::string> deps_chain;
+	ModSpec *spec;
```
Comment (sfan5) [inline src/content/mods.cpp:369]: const?

**Comment by sfan5** [comment]
How hard would it be to add unittests to test for the behaviour that is now no longer broken or other edge cases?

**Comment by appgurueu** [comment]
Here's how a depth-first search may look: https://github.com/appgurueu/modlib/blob/master/minetest/misc.lua#L302-L334

File: src/content/mods.cpp [hunk 4]
```diff
@@ -360,51 +360,127 @@ void ModConfiguration::checkConflictsAndDeps()
...
 			} else {
-				++it;
+				// One of the following cases occured:
+				// 1) Hard-dependency is missing
+				// 2) Recursive dependencies were detected
+
+				// Mod cannot be satisfied: Add to the "error list"
+				// also skip it in the next loop
+				unsatisfied.push_back(mod);
+				mods_by_name.erase(mod_it++);
```
Comment (sfan5) [inline src/content/mods.cpp:461]: I think this must be `mod_it = mods_by_name.erase(mod_it);`

File: src/content/mods.cpp [hunk 5]
```diff
@@ -360,51 +360,127 @@ void ModConfiguration::checkConflictsAndDeps()
...
+}
+
+void ModConfiguration::resolveDependencies()
+{
+	std::unordered_map<std::string, PrioritySortedMod> mods_by_name;
+	for (ModSpec &mod : m_unsatisfied_mods) {
+		// Reduce memory copy time by passing 'mod' as pointer
+		// 'm_unsatisfied_mods' owns the value!
+		mods_by_name.emplace(std::pair<std::string, PrioritySortedMod>(
+				mod.name, &mod));
```
Comment (sfan5) [inline src/content/mods.cpp:432]:
```suggestion
		mods_by_name[mod.name] = PrioritySortedMod(&mod);
```

File: src/unittest/test_servermodmanager.cpp [hunk 6]
```diff
@@ -191,3 +193,84 @@ void TestServerModManager::testGetModMediaPaths()
...
+
+	mods.emplace_back(ModSpec("circ_1"));
+	mods[mods.size() - 1].depends.insert("default");
+	mods[mods.size() - 1].optdepends.insert("circ_2");
+
+	mods.emplace_back(ModSpec("circ_2"));
+	mods[mods.size() - 1].depends.insert("circ_3");
+	mods.emplace_back(ModSpec("circ_3"));
+	mods[mods.size() - 1].depends.insert("circ_1");
+
```
Comment (sfan5) [inline src/unittest/test_servermodmanager.cpp:248]: You can use `mods.back()` instead of `mods[mods.size() - 1]` everywhere here.

**Review by sfan5 — COMMENTED** [review]
Appears to work other than that.

File: src/content/mods.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/content/mods.cpp]: Indexing maps does implicitly call the underlying `PrioritySortedMod` constructor with zero arguments. Worked around this issue by using the default value `nullptr`.

File: src/content/mods.cpp (hunk 4 above)
Comment (SmallJoker) [reply src/content/mods.cpp]: There's a few places with unordered maps which use this notation, namely `m_playing_sounds` and `m_particle_spawners` across the code. I will adjust this to ease readability.

File: src/unittest/test_servermodmanager.cpp (hunk 6 above)
Comment (SmallJoker) [reply src/unittest/test_servermodmanager.cpp:248]: (Done)

File: src/content/mods.cpp (hunk 4 above)
Comment (sfan5) [reply src/content/mods.cpp]: It's not a readability thing. [`erase()`](https://en.cppreference.com/w/cpp/container/unordered_map/erase) is specified to return the element following the removed element.
Because of iterator invalidation I'm not sure if `container.erase(it); it++;` is guaranteed to work.

File: src/content/mods.cpp (hunk 5 above)
Comment (sfan5) [reply src/content/mods.cpp]: Ahhh, right. This should work then even without a default constructor:
```diff
-		mods_by_name.emplace(std::pair<std::string, PrioritySortedMod>(
-				mod.name, &mod));
+		mods_by_name.emplace(mod.name, &mod);
```

If not that's also okay, doesn't really matter.

**Review by sfan5 — APPROVED** [review]
LGTM

**Comment by appgurueu** [comment]
>     2. Each mod adds its own name recursively to the dependency list of each dependent mod

This looks like a suboptimal implementation of topological ordering. Consider a long chain of mods depending on each other - you will already be using quadratic space for your dependency lists.

A proper implementation should run in linear time and space in the size of the dependencies + the mod count (edges / vertices in the dependency graph). Arguably, quadratic time and space is probably fine for rather small mod counts; as this improvement intends to handle rather pathological cases though, I'd say linear time is a must.

**Comment by rubenwardy** [comment]
See #12217
